oci/oci.go

```diff
@@ -514,20 +514,34 @@ func (r *Runtime) StopContainer(c *Container, timeout int64) error {
 		timeout = 10
 	}
 	done := make(chan struct{})
+	// we could potentially re-use "done" channel to exit the loop on timeout
+	// but we use another channel "chControl" so that we won't never incur in the
+	// case the "done" channel is closed in the "default" select case and we also
+	// reach the timeout in the select below. If that happens we could raise
+	// a panic closing a closed channel so better be safe and use another new
+	// channel just to control the loop.
+	chControl := make(chan struct{})
 	go func() {
 		for {
-			// Check if the process is still around
-			err := unix.Kill(c.state.Pid, 0)
-			if err == syscall.ESRCH {
-				close(done)
-				break
+			select {
+			case <-chControl:
+				return
+			default:
+				// Check if the process is still around
+				err := unix.Kill(c.state.Pid, 0)
+				if err == syscall.ESRCH {
+					close(done)
+					return
+				}
+				time.Sleep(100 * time.Millisecond)
 			}
```

Comment:
> Repeat the above, I don't see how that path escapes unless you eventually get syscall.ESRCH.

You're missing the timeout expiring at line 543: when that's hit, we call `close(chControl)`. After that's closed, we don't hit the `default` case again in the select above, rather we hit `case <-chControl` which just returns and exit the loop. That's a golang common idiom and how this is handled. You don't always enter the `default` as you pointed out above.

read my previous comment: we can't get stuck here at all because when the timeout is hit at line 544, we close the chControl (close(chcontrol)). After that's closed the select at line 527 will enter case <-chControl: and do a return exiting the loop and goroutine.